pkg/apis/nfd/validate/validate.go

```diff
@@ -0,0 +1,121 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validate
+
+import (
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	k8svalidation "k8s.io/apimachinery/pkg/util/validation"
+
+	nfdv1alpha1 "sigs.k8s.io/node-feature-discovery/pkg/apis/nfd/v1alpha1"
+)
+
+var (
+	// Default error message for invalid label/annotation keys
+	ErrNSNotAllowed = fmt.Errorf("namespace is not allowed")
+)
+
+// Label validates a label key and value and returns an error if the key or
+// value is invalid.
+func Label(key, value string, features *nfdv1alpha1.Features) error {
+
+	//Validate label name
+	if errs := k8svalidation.IsQualifiedName(key); len(errs) > 0 {
+		return fmt.Errorf("invalid name %q: %s", key, strings.Join(errs, "; "))
+	}
+
+	// Check label namespace, filter out if ns is not whitelisted
+	ns, _ := splitNs(key)
+	if ns != nfdv1alpha1.FeatureLabelNs && ns != nfdv1alpha1.ProfileLabelNs &&
+		!strings.HasSuffix(ns, nfdv1alpha1.FeatureLabelSubNsSuffix) && !strings.HasSuffix(ns, nfdv1alpha1.ProfileLabelSubNsSuffix) {
+		return ErrNSNotAllowed
+	}
+
+	return nil
+}
+
+// Annotation validates an annotation key and value and returns an error if the
+// key or value is invalid.
+func Annotation(key, value string) (string, error) {
+	ns, _ := splitNs(key)
+	if ns != nfdv1alpha1.FeatureAnnotationNs && !strings.HasSuffix(ns, nfdv1alpha1.FeatureAnnotationSubNsSuffix) {
+		// If the namespace is denied, rule is invalid
+		if ns == "kubernetes.io" || strings.HasSuffix(ns, ".kubernetes.io") {
+			return "", fmt.Errorf("namespace %v is not allowed", ns)
+		}
+	}
+	// Validate the annotation value
+	if errs := k8svalidation.IsValidLabelValue(value); len(errs) > 0 {
+		return "", fmt.Errorf("invalid value %q: %s", value, strings.Join(errs, "; "))
+	}
+
+	return value, nil
+}
+
+// Taint validates a taint key and value and returns an error if the key or
+// value is invalid.
+func Taint(taint *corev1.Taint) error {
+	ns, _ := splitNs(taint.Key)
+	// if ns is empty, it means that the rule is not allowed
+	if ns == "" {
+		return fmt.Errorf("unprefixed keys are disallowed")
+	}
+	if ns != nfdv1alpha1.TaintNs && !strings.HasSuffix(ns, nfdv1alpha1.TaintSubNsSuffix) {
+		// If the namespace is denied, rule is invalid
+		if ns == "kubernetes.io" || strings.HasSuffix(ns, ".kubernetes.io") {
+			return fmt.Errorf("namespace %v is not allowed", ns)
+		}
+	}
+	// Validate taint effect is not empty
+	if taint.Effect == "" {
+		return fmt.Errorf("taint effect cannot be empty")
+	}
+	// Validate effect to be only one of NoSchedule, PreferNoSchedule or NoExecute
+	if taint.Effect != "NoSchedule" && taint.Effect != "PreferNoSchedule" && taint.Effect != "NoExecute" {
```

Comment:
I think we can use taints that are already defined in the `corev1` package of Kubernetes. e.g. `taint.Effect != corev1.TaintEffectNoSchedule`